ENH add D2 pinbal score and D2 absolute error score (#22118)
scikit-learn/scikit-learn: doc/modules/model_evaluation.rst

'neg_mean_poisson_deviance'            :func:`metrics.mean_poisson_deviance`
'neg_mean_gamma_deviance'              :func:`metrics.mean_gamma_deviance`
'neg_mean_absolute_percentage_error'   :func:`metrics.mean_absolute_percentage_error`
'd2_tweedie_score'                     :func:`metrics.d2_tweedie_score`
'd2_pinball_score'                     :func:`metrics.d2_pinball_score`

> 'd2_absolute_error_score'              :func:`metrics.d2_absolute_error_score`

Reviewer: Please revert the order such that it is sorted.

we would get identical errors. The deviance when ``power=2`` is thus only
sensitive to relative errors.

.. _d2_score:

D² score
--------

The D² score computes the fraction of deviance explained. 
It is a generalization of R², where the squared error is generalized and replaced 

> by a deviance of choice :math:`\text{D}(y, \hat{y})`

Reviewer: Maybe, we should avoid confusion with D2 and name deviance with a small `\text{dev}` instead of `\text{D}`.

Author: Good idea, thanks! updated.

(e.g., Tweedie, pinball or mean absolute error).
It is calculated as

.. math::

>   D^2(y, \hat{y}) = 1 - \frac{\text{D}(y, \hat{y})}{\text{D}(y, \bar{y})} \,.

Reviewer: D^2(y, \hat{y}) = 1 - \frac{\text{D}(y, \hat{y})}{\text{D}(y, y_{\text{null}})} \,,

> Where :math:`\bar{y}` is the optimal intercept-only model's prediction 
> (e.g., the mean of `y_true` for the Tweedie case, the median for absolute 
> error and the alpha-quantile for pinball loss).

Reviewer: ```suggestion
Where :math:`y_{\text{null}}` is the optimal intercept-only model's prediction 
(e.g., the mean of `y_true` for the Tweedie case, the median for absolute 
error and the alpha-quantile for pinball loss).
```

Author: Fixed as well

Like R², the best possible score is 1.0 and it can be negative (because the
model can be arbitrarily worse). A constant model that always predicts the
expected value of y, disregarding the input features, would get a D² score
of 0.0.

D² Tweedie score
^^^^^^^^^^^^^^^^

The :func:`d2_tweedie_score` function implements the special case of D² 

> where :math:`\text{D}(y, \hat{y})` is the Tweedie deviance. 

Reviewer: ```suggestion
where :math:`\text{dev}(y, \hat{y})` is the Tweedie deviance. 
```

Author: Fixed

It is also known as D² Tweedie or McFadden's likelihood ratio index.

The argument ``power`` defines the Tweedie power as for
:func:`mean_tweedie_deviance`. Note that for `power=0`,
:func:`d2_tweedie_score` equals :func:`r2_score` (for single targets).

A scorer object with a specific choice of ``power`` can be built by::

  >>> from sklearn.metrics import d2_tweedie_score, make_scorer
  >>> d2_tweedie_score_15 = make_scorer(d2_tweedie_score, power=1.5)

D² pinball score
^^^^^^^^^^^^^^^^^^^^^

The :func:`d2_pinball_score` function implements the special case
of D² with a pinball deviance, i.e.:

.. math::

  D(y, \hat{y}) = \sum_{i=0}^{n_{\text{samples}}-1}  \alpha \max(y_i - \hat{y}_i, 0) + (1 - \alpha) \max(\hat{y}_i - y_i, 0).

> The argument ``alpha`` defines the slope of the pinball deviance as for

Reviewer: Please use "pinball loss" instead of "pinball deviance".

Author: Thanks, fixed!

expected value of y, disregarding the input features, would get a D² score
of 0.0.

D² Tweedie score
^^^^^^^^^^^^^^^^

The :func:`d2_tweedie_score` function implements the special case of D² 
where :math:`\text{dev}(y, \hat{y})` is the Tweedie deviance. 
It is also known as D² Tweedie or McFadden's likelihood ratio index.

The argument ``power`` defines the Tweedie power as for
:func:`mean_tweedie_deviance`. Note that for `power=0`,
:func:`d2_tweedie_score` equals :func:`r2_score` (for single targets).

A scorer object with a specific choice of ``power`` can be built by::

  >>> from sklearn.metrics import d2_tweedie_score, make_scorer
  >>> d2_tweedie_score_15 = make_scorer(d2_tweedie_score, power=1.5)

D² pinball score
^^^^^^^^^^^^^^^^^^^^^

The :func:`d2_pinball_score` function implements the special case
of D² with a pinball deviance, i.e.:

.. math::

  \text{dev}(y, \hat{y}) = \sum_{i=0}^{n_{\text{samples}}-1}  \alpha \max(y_i - \hat{y}_i, 0) + (1 - \alpha) \max(\hat{y}_i - y_i, 0).

The argument ``alpha`` defines the slope of the pinball loss as for
:func:`mean_pinball_loss`. It determines the quantile level ``alpha``
for which the pinball deviance and also D2 are optimal.

> Note that for `alpha=0.5` (the defualt) :func:`d2_pinball_score`

Author: Thanks for commenting @agramfort!
Thanks for catching my typo, it is fixed now. I added a cross-ref to the pinball loss section and also changed the order such that the pinball loss section comes before the D2 section, it definitely makes more sense that way :smile:.
Please take another look and let me know if any more changes are needed, thanks!

    hyper-parameters of quantile regression models on data with non-symmetric
    noise and outliers.

.. _d2_score:

D² score
--------

The D² score computes the fraction of deviance explained. 
It is a generalization of R², where the squared error is generalized and replaced 
by a deviance of choice :math:`\text{dev}(y, \hat{y})`
(e.g., Tweedie, pinball or mean absolute error).
It is calculated as

.. math::

  D^2(y, \hat{y}) = 1 - \frac{\text{dev}(y, \hat{y})}{\text{dev}(y, y_{\text{null}})} \,.

Where :math:`y_{\text{null}}` is the optimal intercept-only model's prediction 
(e.g., the mean of `y_true` for the Tweedie case, the median for absolute 
error and the alpha-quantile for pinball loss).

Like R², the best possible score is 1.0 and it can be negative (because the
model can be arbitrarily worse). A constant model that always predicts the
expected value of y, disregarding the input features, would get a D² score

> of 0.0.

Author: Thanks, Fixed!

D² Tweedie score
^^^^^^^^^^^^^^^^

The :func:`d2_tweedie_score` function implements the special case of D² 

> where :math:`\text{dev}(y, \hat{y})` is the Tweedie deviance. 

Author: Add cross-ref

> It is also known as D² Tweedie or McFadden's likelihood ratio index.

Author: Thanks for catching that, fixed!

The argument ``power`` defines the Tweedie power as for
:func:`mean_tweedie_deviance`. Note that for `power=0`,
:func:`d2_tweedie_score` equals :func:`r2_score` (for single targets).

A scorer object with a specific choice of ``power`` can be built by::

  >>> from sklearn.metrics import d2_tweedie_score, make_scorer
  >>> d2_tweedie_score_15 = make_scorer(d2_tweedie_score, power=1.5)

D² pinball score
^^^^^^^^^^^^^^^^^^^^^

The :func:`d2_pinball_score` function implements the special case

> of D² with a pinball deviance, i.e.:

Author: Corrected now.

.. math::

>   \text{dev}(y, \hat{y}) = \sum_{i=0}^{n_{\text{samples}}-1}  \alpha \max(y_i - \hat{y}_i, 0) + (1 - \alpha) \max(\hat{y}_i - y_i, 0).

Author: Fixed

The argument ``alpha`` defines the slope of the pinball loss as for
:func:`mean_pinball_loss` (:ref:`pinball_loss`). It determines the 

> quantile level ``alpha`` for which the pinball deviance and also D2 

Author: Fixed

are optimal. Note that for `alpha=0.5` (the default) :func:`d2_pinball_score`
equals :func:`d2_absolute_error_score`.

A scorer object with a specific choice of ``alpha`` can be built by::

  >>> from sklearn.metrics import d2_pinball_score, make_scorer
  >>> d2_pinball_score_08 = make_scorer(d2_pinball_score, alpha=0.8)

D² absolute error score
^^^^^^^^^^^^^^^^^^^^^^^

> The :func:`d2_absolute_error_score` function implements the special case of

Reviewer: ```suggestion
The :func:`d2_absolute_error_score` function implements the special case of the :ref:`mean_absolute_error`
```

Author: Added cross-ref

.. math::

>   \text{dev}(y, \hat{y}) = \sum_{i=0}^{n_{\text{samples}}-1} \left| y_i - \hat{y}_i \right|.

Reviewer: ```suggestion
  \text{dev}(y, \hat{y}) = \text{MAE}(y, \hat{y}).
```

Author: Fixed

    hyper-parameters of quantile regression models on data with non-symmetric
    noise and outliers.

.. _d2_score:

D² score
--------

The D² score computes the fraction of deviance explained. 
It is a generalization of R², where the squared error is generalized and replaced 

> by a deviance of choice :math:`\text{dev}(y, \hat{y})`
> (e.g., Tweedie, pinball or mean absolute error).

Reviewer: ```suggestion
by a deviance of choice :math:`\text{dev}(y, \hat{y})` (e.g., Tweedie, pinball
or mean absolute error). D² is a form of a *skill score*.
```

It is calculated as

.. math::

  D^2(y, \hat{y}) = 1 - \frac{\text{dev}(y, \hat{y})}{\text{dev}(y, y_{\text{null}})} \,.

> Where :math:`y_{\text{null}}` is the optimal intercept-only model's prediction 

Reviewer: ```suggestion
Where :math:`y_{\text{null}}` is the optimal prediction of an intercept-only model
```
